src/box/lua/net_box.c

```diff
@@ -3080,6 +3080,25 @@ netbox_connection_handler_f(struct lua_State *L)
 	}
 	return 0;
 }
+
+/**
+ * A wrapper around `fiber_is_cancelled` that closes the transport.
+ */
+static bool
+netbox_worker_is_cancelled(struct netbox_transport *transport,
```

Comment:
1. It's unexpected that a function named `netbox_worker_is_cancelled` can actually change the object state.
2. Clearing `is_closing` here looks redundant.
3. Joining the worker fiber is unsafe because `close` may be called concurrently from different fibers. Calling `fiber_join` on the same fiber more than once is prohibited.
4. Passing the extra flag `in_finalizer` to `luaT_netbox_transport_stop` looks ugly. Instead we can reuse the flag `wait`.

<details><summary>Suggested patch</summary>

```diff
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index 8b0f693e2806..a7cf226c0888 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -252,6 +252,8 @@ struct netbox_transport {
 	struct error *last_error;
 	/** Fiber doing I/O and dispatching responses. */
 	struct fiber *worker;
+	/** Signalled when the worker is stopped. */
+	struct fiber_cond on_worker_stop;
 	/**
 	 * Lua reference to the Lua state used by the worker fiber or
 	 * LUA_NOREF if the worker fiber isn't running.
@@ -544,6 +546,7 @@ netbox_transport_create(struct netbox_transport *transport)
 	transport->is_closing = false;
 	transport->last_error = NULL;
 	transport->worker = NULL;
+	fiber_cond_create(&transport->on_worker_stop);
 	transport->coro_ref = LUA_NOREF;
 	transport->self_ref = LUA_NOREF;
 	iostream_ctx_clear(&transport->io_ctx);
@@ -566,6 +569,7 @@ netbox_transport_destroy(struct netbox_transport *transport)
 	assert(transport->worker == NULL);
 	assert(transport->coro_ref == LUA_NOREF);
 	assert(transport->self_ref == LUA_NOREF);
+	fiber_cond_destroy(&transport->on_worker_stop);
 	iostream_ctx_destroy(&transport->io_ctx);
 	assert(!iostream_is_initialized(&transport->io));
 	assert(ibuf_used(&transport->send_buf) == 0);
@@ -3103,7 +3107,7 @@ netbox_worker_f(va_list ap)
 			(void)rc;
 			iostream_close(&transport->io);
 		}
-		if (transport->state == NETBOX_CLOSED)
+		if (fiber_is_cancelled())
 			break;
 		netbox_transport_set_error(transport);
 		transport->state = (reconnect_after > 0 ?
@@ -3115,7 +3119,14 @@ netbox_worker_f(va_list ap)
 			break;
 		}
 	}
+	if (transport->state != NETBOX_ERROR) {
+		box_error_raise(ER_NO_CONNECTION, "Connection closed");
+		netbox_transport_set_error(transport);
+		transport->state = NETBOX_CLOSED;
+		netbox_transport_on_state_change_pcall(transport, L);
+	}
 	transport->worker = NULL;
+	fiber_cond_broadcast(&transport->on_worker_stop);
 	luaL_unref(tarantool_L, LUA_REGISTRYINDEX, transport->coro_ref);
 	transport->coro_ref = LUA_NOREF;
 	/* Careful: luaL_unref may delete this transport object. */
@@ -3171,17 +3182,17 @@ luaT_netbox_transport_start(struct lua_State *L)
  * Stops a worker fiber.
  *
  * Takes an optional boolean argument 'wait': if set the function will wait
- * for all pending requests to be sent.
+ * for all pending requests to be sent and for the connection to be closed.
  */
 static int
 luaT_netbox_transport_stop(struct lua_State *L)
 {
 	struct netbox_transport *transport = luaT_check_netbox_transport(L, 1);
+	transport->is_closing = true;
 	bool wait = lua_toboolean(L, 2);
-	if (wait && fiber() != transport->worker &&
-	    transport->state != NETBOX_CLOSED &&
-	    transport->state != NETBOX_ERROR) {
-		transport->is_closing = true;
+	if (fiber() == transport->worker)
+		wait = false;
+	if (wait) {
 		/*
 		 * Here we are waiting until send buf became empty:
 		 * it is necessary to ensure that all requests are
@@ -3189,18 +3200,6 @@ luaT_netbox_transport_stop(struct lua_State *L)
 		 */
 		while (ibuf_used(&transport->send_buf) > 0)
 			fiber_cond_wait(&transport->on_send_buf_empty);
-		transport->is_closing = false;
-	}
-	/*
-	 * While we were waiting for the send buffer to be empty,
-	 * the state could change.
-	 */
-	if (transport->state != NETBOX_CLOSED &&
-	    transport->state != NETBOX_ERROR) {
-		box_error_raise(ER_NO_CONNECTION, "Connection closed");
-		netbox_transport_set_error(transport);
-		transport->state = NETBOX_CLOSED;
-		netbox_transport_on_state_change(transport, L);
 	}
 	/* Cancel the worker fiber. */
 	if (transport->worker != NULL) {
@@ -3208,6 +3207,11 @@ luaT_netbox_transport_stop(struct lua_State *L)
 		/* Check if we cancelled ourselves. */
 		luaL_testcancel(L);
 	}
+	if (wait) {
+		/* Wait for the worker fiber to stop. */
+		while (transport->worker != NULL)
+			fiber_cond_wait(&transport->on_worker_stop);
+	}
 	return 0;
 }
 
```

</details>

Reply:
> 3. Joining the worker fiber is unsafe because `close` may be called concurrently from different fibers. Calling `fiber_join` on the same fiber more than once is prohibited.

Thanks for the catch, I didn't know about this — added a test case to `box-luatest/net_box_test.lua` for this.

Applied your patch, I don't see any objections.